lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
+private:
+  /// The global context.
+  ADContext &context;
+
+  /// The original function.
+  SILFunction *const original;
+
+  /// The `[differentiable]` attribute.
+  SILDifferentiableAttr *const attr;
+
+  /// The JVP function.
+  SILFunction *const jvp;
+
+  /// The differential function.
+  SILFunction *differential;
+
+  /// The differentiation invoker.
+  DifferentiationInvoker invoker;
+
+  /// Caches basic blocks whose phi arguments have been remapped (adding a
+  /// predecessor enum argument).
+  SmallPtrSet<SILBasicBlock *, 4> remappedBasicBlocks;
+
+  /// Temporary boolean for determine whether we should be generating the body of the JVP.
+  /// If true, then skip the JVP generation process since we cannot generate JVPs for the
+  /// primitive VJPs defined on addition for Float for example.
+  bool vjpGenerated;
+
+  bool errorOccurred = false;
+
+  ASTContext &getASTContext() const { return jvp->getASTContext(); }
+  SILModule &getModule() const { return jvp->getModule(); }
+  const SILAutoDiffIndices &getIndices() const { return attr->getIndices(); }
+
+  static SubstitutionMap getSubstitutionMap(SILFunction *original,
+                                            SILFunction *jvp) {
+    auto substMap = original->getForwardingSubstitutionMap();
+    if (auto *jvpGenEnv = jvp->getGenericEnvironment())
+      substMap = substMap.subst(jvpGenEnv->getForwardingSubstitutionMap());
+    return substMap;
+  }
+
+public:
+  explicit JVPEmitter(ADContext &context, SILFunction *original,
+                      SILDifferentiableAttr *attr, SILFunction *jvp,
+                      DifferentiationInvoker invoker, bool vjpGenerated)
+      : TypeSubstCloner(*jvp, *original, getSubstitutionMap(original, jvp)),
+        context(context), original(original), attr(attr), jvp(jvp),
+        invoker(invoker),
+        vjpGenerated(vjpGenerated) {
+    // Create empty differential function.
+    differential = createEmptyDifferential();
+    context.getGeneratedFunctions().push_back(differential);
+  }
+
+  SILFunction *createEmptyDifferential() {
+    auto &module = context.getModule();
+    auto origTy = original->getLoweredFunctionType();
+    auto lookupConformance = LookUpConformanceInModule(module.getSwiftModule());
+
+    // RAII that pushes the original function's generic signature to
+    // `module.Types` so that the calls `module.Types.getTypeLowering()` below
+    // will know the original function's generic parameter types.
+    Lowering::GenericContextScope genericContextScope(
+        module.Types, origTy->getGenericSignature());
+
+    SmallVector<SILParameterInfo, 8> diffParams;
+    SmallVector<SILResultInfo, 8> diffResults;
+    auto origParams = origTy->getParameters();
+    auto indices = attr->getIndices();
+
+    // Add differential result for the seed.
+    auto origResInfo = origTy->getResults()[indices.source];
+    diffResults.push_back(SILResultInfo(
+                              origResInfo.getType()
```

Comment:
I feel the indentation is very awkward here. The convention we follow in the AD code base is usually indenting each line break by exactly 4 regardless of the parent expression's start position.

Reply:
Fair, I will put
```swift
diffResults.push_back(SILResultInfo(origResInfo.getType()
        ->getAutoDiffAssociatedTangentSpace(lookupConformance)
        ->getCanonicalType(), origResInfo.getConvention()));
```